### ISSUE TITLE
Does SSDP discovery still work?

### ISSUE THREAD

**Issue by erodewald** [R0nd/LgSmartTvRemote#1]
TV model: 2013 LG *60PN5700-UA*.

I am attempting to discover the device using your library which depends on `node-ssdp`, which does not set `USER-AGENT` when it sends the `m-search * HTTP1.1` request. I do get a "pairing dialog" that pops up on the screen, but it has no pairing code. 

According to their documentation, `USER-AGENT` is required by UDAP (which is built on top of UPnP, which has it as an optional field when using SSDP m-search). See here: http://developer.lgappstv.com/TV_HELP/index.jsp?topic=%2Flge.tvsdk.references.book%2Fhtml%2FUDAP%2FUDAP%2FLG+UDAP+2+0+Protocol+Specifications.htm

I'm not sure if this ever really worked, or if they updated it in a later software update.

**Comment by R0nd** [comment]
What's your OS? For some reason node-ssdp doesn't seem to work at all on Windows 10, but works fine on Linux. Looking into it.

**Comment by R0nd** [comment]
Using explicit interface binding in SSDP client seems to have fixed discovery on Windows.
I have published the package v1.0.1 with this change.

> I do get a "pairing dialog" that pops up on the screen, but it has no pairing code.

I believe this dialog pops up on discovery, after which you're supposed to request the pairing code with
`lgremote.requestPairingKey();
`

**Comment by erodewald** [comment]
Sorry to keep you hanging. I was using it on macOS High Sierra. For what it's worth, it found _several_ devices on my LAN with that particular `urn` -- about 8 for my Hue hub (I'm guessing it registers itself for different capabilities?). 

I was calling `requestPairingKey()` on success for discovery, but it never got that far because all of the discovered devices didn't seem to match the LG TV. 

Of course, it works to just know the device's IP, but in my scenario it would be highly advantageous to using device discovery. 

Let me know if you figure anything out, but unfortunately there isn't much I can do with the LG TV's API.

**Comment by R0nd** [comment]
Yep, it counts any device that responds to "udap:rootservice" as a valid LG TV, that would lead to discovery dialog popping up on the TV and the pairing request going to some other device. Gotta find a way to distinguish the discovery response from TV. I will take a look at this.

**Comment by erodewald** [comment]
There's a couple ways to do that. Hook into the SSDP server's `message` event:
1. parse out the XML for a manufacturer name of `LG Electronics`
2. Don't parse, just check if you see the string that would indicate a compatible TV: 
    `if (message.indexOf("LG Smart TV"))...`

I [found an example](http://developer.lgappstv.com/TV_HELP/index.jsp?topic=%2Flge.tvsdk.references.book%2Fhtml%2FUDAP%2FUDAP%2FDiscovery.htm) of how the LG TV will identify itself.

```xml
<?xml version="1.0" encoding="utf-8"?>
<envelope>
    <device>
        <deviceType>type of Host device</deviceType>
        <modelName>model name of Host device</modelName>
        <friendlyName>user-friendly name of a model</friendlyName>
        <manufacturer>LG Electronics</manufacturer>
        <uuid>Host device UUID value</uuid>
    </device>
</envelope>
```

If you have more than 1 LG Netcast TV in the house, this might be unreliable. You could simply trigger pairing to all returned device IPs. 

Hope that helps.

**Comment by R0nd** [comment]
Okay, I've reworked the discovery process:
1. Now it uses the concrete `urn:schemas-udap:service:netrcu:1` UDAP service instead of querying any root service
2. For every responded device, the service description XML is checked to have deviceType === "TV" and manufacturer === "LG Electronics"
3. For the case when there are more than one TVs on the network, I've added an optional `deviceName` argument to the discover function, which will be used to match the device's `friendlyName`

Please update the package to v1.0.2 and let me know if that works.

**Comment by R0nd** [comment]
I've published v1.0.3 with final changes, that should work now. Also added some Typescript typings.

**Comment by erodewald** [comment]
It does seem to work now. There are quirks that I'm attributing to the TV, like sometimes it completely doesn't respond to the SSDP discovery even though the "popup" appears. Wait a bit, retry, and it suddenly works. 🤕 

It's probably worth mentioning that although it's using the more precise `urn:schemas-udap:service:netrcu:1` service, my Hue Hub still response. It would seem it's not exclusive to LG. This means that I get some exceptions because it's response to the `getDescription()` query is differently-structured to the LG response. 

I have a pull request to fix it in a safe fashion. Since you might not have anything to test it out, I can confirm it just passes right by the conditions on [index.js:L109-L113](https://github.com/R0nd/LgSmartTvRemote/blob/master/index.js#L109-L113) as the value becomes `undefined`.
